### PR TITLE
[WIP] -MFCS- Modular Firearm Construction System

### PR DESCRIPTION
**Pull request by BlueNexus**
//THIS IS HEAVILY WORK IN PROGRESS//

The MFCS allows players to construct and disassemble firearms using a wide variety of components, allowing the player to customise recoil, accuracy, projectile type, fire rate and mode, projectile speed, loading method and more. All firearms in the game will be replaced with these modular variants. MFCS Firearms may be constructed by R&D or anybody who gets a hold of the parts, however the final part will come in a lock box, to prevent them being made and used for grief. This will hopefully give research something interesting to do, and will add a lot of depth to the firearms in the game. Keep in mind, this project is massively work in progress. Feedback would be appreciated, however.

Current build - 215

### CODE REVIEW

**Comment by Lorwp** [comment]
Is this in a state where it could be tested?

**Comment by BlueNexus** [comment]
@Lorwp if you mean does it compile, it does. However, all that's finished is the first two steps of construction, and there's currently a fatal infinite loop error when attempting the second step.

**Comment by Lorwp** [comment]
@BlueNexus Got it, keep us posted on this, it looks interesting :+1:

**Comment by BlueNexus** [comment]
Update - Stage four of the assembly is complete. 
Drivers = Done
Loaders = Done
Chassis = Sort of done
Barrels = Not started
Chambers = Done
Accessories = Not started

File: code/modules/projectiles/guns/modular/assembly.dm [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+			buildstage += 1
+		else
+			user << "\red You must install a driver first!"
+	else if(buildstage == 4)
+		if(istype(I, /obj/item/weapon/modular_firearms/loader))
+			user.drop_item()
+			I.loc = src
+			components += I
+			modLoader = I
+			var/obj.item/weapon/modular_firearms/loader/L = I
```
Comment (faerindel) [inline code/modules/projectiles/guns/modular/assembly.dm]: Typo? Or Byond swallows that var/obj.item?

**Comment by BlueNexus** [comment]
@faerindel _sighs_ Typo. Thanks for pointing it out.

**Comment by BlueNexus** [comment]
Cleaned it up to massively reduce the amount of code needed. Adding parts and processing them is now handled by a single proc and a sub-proc. Will finish the current parts, then work on deconstruction. Finally, processing the gun itself.

**Pull request closed without merging**